src/posix/platform/settings.cpp

```diff
@@ -32,429 +32,443 @@
+        rval = read(sSettingsFd, &key, sizeof(key));
+        VerifyOrExit(rval == sizeof(key), error = OT_ERROR_PARSE);
 
-    initSettings(sSettingsBaseAddress, static_cast<uint32_t>(kSettingsInSwap));
-    sSettingsUsedSize = kSettingsFlagSize;
-    swapAddress += kSettingsFlagSize;
+        rval = read(sSettingsFd, &length, sizeof(length));
+        VerifyOrExit(rval == sizeof(length), error = OT_ERROR_PARSE);
+
+        offset += sizeof(key) + sizeof(length) + length;
+        VerifyOrExit(offset == lseek(sSettingsFd, length, SEEK_CUR), error = OT_ERROR_PARSE);
+    }
 
-    while (swapAddress < (oldBase + usedSize))
+exit:
+    if (error == OT_ERROR_PARSE)
     {
-        OT_TOOL_PACKED_BEGIN
-        struct addSettingsBlock
-        {
-            struct settingsBlock block;
-            uint8_t              data[kSettingsBlockDataSize];
-        } OT_TOOL_PACKED_END addBlock;
-        bool                 valid = true;
+        VerifyOrDie(ftruncate(sSettingsFd, 0) == 0);
+    }
+}
 
-        utilsFlashRead(swapAddress, reinterpret_cast<uint8_t *>(&addBlock.block), sizeof(struct settingsBlock));
-        swapAddress += sizeof(struct settingsBlock);
+otError otPlatSettingsGet(otInstance *aInstance, uint16_t aKey, int aIndex, uint8_t *aValue, uint16_t *aValueLength)
+{
+    otError     error  = OT_ERROR_NOT_FOUND;
+    const off_t size   = lseek(sSettingsFd, 0, SEEK_END);
+    off_t       offset = lseek(sSettingsFd, 0, SEEK_SET);
 
-        if (!(addBlock.block.flag & kBlockAddCompleteFlag) && (addBlock.block.flag & kBlockDeleteFlag))
-        {
-            uint32_t address = swapAddress + getAlignLength(addBlock.block.length);
+    OT_UNUSED_VARIABLE(aInstance);
+    VerifyOrExit(offset == 0 && size >= 0, error = OT_ERROR_PARSE);
 
-            while (address < (oldBase + usedSize))
-            {
-                struct settingsBlock block;
+    while (offset < size)
+    {
+        uint16_t key;
+        uint16_t length;
+        ssize_t  rval;
 
-                utilsFlashRead(address, reinterpret_cast<uint8_t *>(&block), sizeof(block));
+        rval = read(sSettingsFd, &key, sizeof(key));
+        VerifyOrExit(rval == sizeof(key), error = OT_ERROR_PARSE);
 
-                if (!(block.flag & kBlockAddCompleteFlag) && (block.flag & kBlockDeleteFlag) &&
-                    !(block.flag & kBlockIndex0Flag) && (block.key == addBlock.block.key))
+        rval = read(sSettingsFd, &length, sizeof(length));
+        VerifyOrExit(rval == sizeof(length), error = OT_ERROR_PARSE);
+
+        if (key == aKey)
+        {
+            if (aIndex == 0)
+            {
+                error = OT_ERROR_NONE;
+
+                if (aValueLength)
                 {
-                    valid = false;
-                    break;
+                    if (aValue)
+                    {
+                        VerifyOrExit(read(sSettingsFd, aValue, (length <= *aValueLength ? length : *aValueLength)) ==
+                                         length,
```

Comment:
Thanks for changing this.
 `== length` seems incorrect in case `length > *aValueLength` and we read `*aValueLength` bytes?